[new] promtail pipeline:  Promtail Rate Limit stage #5048 (#5051)
grafana/loki: docs/sources/clients/promtail/stages/limit.md

---
title: limit
---
# `limit` stage

> The `limit` stage is a ratelimit stage that lets you throttle logs based on several options. 

Reviewer: ```suggestion
The `limit` stage is a rate limiting stage throttles logs based on several options. 
```

## Limit stage schema

```yaml
limit:

>   # Name from extracted data to parse. If empty, uses the log message.
>   [source: <string>]
>   
>   # RE2 regular expression, if source is provided the regex will attempt to match the source
>   # If no source is provided, then the regex attempts to match the log line
>   # If the provided regex matches the log line or a provided source, the line will be throttled. 
>   [expression: <string>]
> 
>   # value can only be specified when source is specified. It is an error to specify value and regex.
>   # If the value provided is an exact match for the `source` the line will be throttled.
>   [value: <string>]
> 
>   # allowed ingestion rate limit in lines per second
>   [rate: <int>]
> 
>   # allowed ingestion burst lines
>   [burst: <int>]
> 
>   # When drop is equal to true, logs that exceed the current ratelimit will be discarded.
>   # When drop is equal to false, the log that exceeds the current ratelimit will only wait to enter the backpressure mode 
>   [drop: <bool> | default = false]

Reviewer: I'm correcting grammar and punctuation. I'm also reformatting such that line lengths will be smaller. This prevents the rendered page from needing a horizontal scroll bar.

```suggestion
  # Name from the extracted data to parse. If empty, it parses the log message.
  [source: <string>]
  
  # An RE2 regular expression. If the source is provided, the regex will attempt to match
  # the source. If no source is provided, then the regex attempts to match the log line.
  # If the provided regex matches the log line or a provided source, the line will
  # be throttled. 
  [expression: <string>]

  # A value can only be specified when `source` is specified. It is an error to specify
  # both `value` and `expression`. If the value provided is an exact match for
  # the `source`, the line will be throttled.
  [value: <string>]

  # The allowed ingestion rate limit in lines per second
  [rate: <int>]

  # The quantity of allowed ingestion burst lines
  [burst: <int>]

  # When drop is true, log lines that exceed the current rate limit will be discarded.
  # When drop is false, log lines that exceed the current rate limit will only wait
  # to enter the back pressure mode. 
  [drop: <bool> | default = false]
```

---
title: limit
---
# `limit` stage

The `limit` stage is a rate limiting stage throttles logs based on several options. 

## Limit stage schema

> 
> ```yaml
> limit:
>   # The allowed ingestion rate limit in lines per second
>   [rate: <int>]
> 
>   # The quantity of allowed ingestion burst lines
>   [burst: <int>]

Reviewer: ```suggestion

This pipeline stage places limits on the rate or burst quantity of log lines that Promtail pushes to Loki.
The rate and burst limits are equivalent to the limits specified by the distributor configuration parameters  `ingestion_rate_mb` and `ingestion_burst_size_mb`, as defined in [limits_config](../../../../configuration/#limits_config).

```yaml
limit:
  # The rate limit in lines per second that Promtail will push to Loki
  [rate: <int>]

  # The cap in the quantity of burst lines that Promtail will push to Loki
  [burst: <int>]
```

Reviewer 2: I made an edit to the comment above so it now reads
```
This pipeline stage places limits on the rate or burst quantity of log lines that Promtail pushes to Loki.
The concept of having distinct burst and rate limits mirrors the approach to limits that can be set for Loki's distributor component:  `ingestion_rate_mb` and `ingestion_burst_size_mb`, as defined in [limits_config](../../../../configuration/#limits_config).

```

  # When drop is true, log lines that exceed the current rate limit will be discarded.
  # When drop is false, log lines that exceed the current rate limit will only wait
  # to enter the back pressure mode. 
  [drop: <bool> | default = false]
```

## Examples

The following are examples showing the use of the `limit` stage.

### limit

Simple `limit` stage configurations.

#### Regex match a line and throttle

Given the pipeline:

```yaml
- limit:
    rate: 10
    burst: 10
```

Would throttle any log line.

> #### Regex match a line and drop

Reviewer 3: ```suggestion
#### Match a line and drop
```
